pyro/infer/mcmc/trace_kernel.py

```diff
@@ -0,0 +1,61 @@
+from __future__ import absolute_import, division, print_function
+
+from abc import ABCMeta, abstractmethod
+
+from six import add_metaclass
+
+
+@add_metaclass(ABCMeta)
+class TraceKernel(object):
+
+    def setup(self, *args, **kwargs):
+        """
+        Optional method to set up any state required before the run.
+
+        :param \*args: Algorithm specific positional arguments.
+        :param \*\*kwargs: Algorithm specific keyword arguments.
+        """
+        pass
+
+    def cleanup(self):
+        """
+        Optional method to clean up any residual state on termination.
+        """
+        pass
+
+    def diagnostics(self, time_step):
+        """
+        Relevant diagnostics (optional) to be printed at regular intervals
+        of the MCMC run. Returns `None` by default.
+
+        :param time_step: Current Monte Carlo time-step.
+        :return: String containing the diagnostic summary. e.g. acceptance rate
+        :rtype: string
+        """
+        return None
+
+    @abstractmethod
+    def initial_trace(self):
+        """
+        Returns an initial trace from the prior to initiate the MCMC run.
+
+        :return: Trace instance.
+        """
+        return NotImplementedError
+
+    @abstractmethod
+    def sample(self, trace):
+        """
+        Samples a trace from the approximate posterior distribution, when given an existing trace.
+
+        :param trace: Current execution trace.
+        :param int time_step: Current time step.
+        :return: New trace sampled from the approximate posterior distribution.
+        """
+        raise NotImplementedError
+
+    def __call__(self, trace):
```

Comment:
Just curious, do you have a reason for spelling this out rather than a one-liner?
```py
__call__ = sample
```

Reply:
Not really. Old habits I guess. :) Will change.

Reply:
Actually, there is an interesting issue that I encountered. If we do this, `__call__` will become an `abstractmethod` too and will need to be explicitly implemented by the sub class. Which makes sense in a way, but makes the abstract class functionality seem like a hack.

```
Can't instantiate abstract class PriorKernel with abstract methods __call__
```

Reply:
Ah good to know, thanks for explaining.